### ISSUE TITLE
could you provide the template applicability filter policy to help speed up the search time?

### ISSUE THREAD

**Issue by queliyong** [MolecularAI/aizynthfinder#52]
Hello,
After reading the paper "Artificial applicability labels for improving policies in retrosynthesis prediction", I am wondering if you could provide the filter policy to help speed up the search time.

**Comment by EBjerrum** [comment]
Thanks for you interest in the approach. The applicability filter do not help to speed up the search time, but do over longer runs increase the number of solved routes somewhat (c.f. Figure 9 from the paper).
We have not come around to train it on the public data, and as it is tightly coupled to the template set, this would be necessary. What policy do you intent to use it for?

**Comment by queliyong** [comment]
> 

Thanks for your patient reply. It is well known that some of the top 50 predicted templates by expansion policy are nonapplicable, and this consumes much of the run time during the search, so if the template applicability filter policy is used to filter the nonapplicable templates, the search must be faster. am I right? thank u.

**Comment by queliyong** [comment]
> Thanks for you interest in the approach. The applicability filter do not help to speed up the search time, but do over longer runs increase the number of solved routes somewhat (c.f. Figure 9 from the paper). We have not come around to train it on the public data, and as it is tightly coupled to the template set, this would be necessary. What policy do you intent to use it for?

Thanks for your patient reply. It is well known that some of the top 50 predicted templates by expansion policy are nonapplicable, and this consumes much of the run time during the search, so if the template applicability filter policy is used to filter the nonapplicable templates, the search must be faster. am I right? thank u. But the template applicability filter policy and its corresponding training data are not easy to get, so i attempt to ask for help.

**Comment by EBjerrum** [comment]
It's not really faster for most of the cases. What it does, is that it gives the search more lower priority options to try. The code that tries and removes nonapplicable templates during search is quite fast, so for the top prioritized templates with the standard policy network it doesn't matter so much speed-wise, if the top-5 applicable predictions are enough to lead to a successful route. On the longer run it may be true that it's faster, you could as an alternative to the applicability filter, simply use the, say,  top 10000 templates from a regular policy, where maybe 50 are useful, and then the time used for eliminating 9950 of the templates explicitly would start to matter. So for longer runs on tougher to predict compounds, it leads to a higher success rate of finding routes as there are more options for the tree search, but for the majority of compounds it does not give faster search times to first route found (see Figure 9 in the MLST paper).
If the lower priority, but applicable, templates are then merely applicable in silico and not chemically relevant, is something that's a bit tougher to measure, but we are working on that aspect.

**Comment by queliyong** [comment]
> It's not really faster for most of the cases. What it does, is that it gives the search more lower priority options to try. The code that tries and removes nonapplicable templates during search is quite fast, so for the top prioritized templates with the standard policy network it doesn't matter so much speed-wise, if the top-5 applicable predictions are enough to lead to a successful route. On the longer run it may be true that it's faster, you could as an alternative to the applicability filter, simply use the, say, top 10000 templates from a regular policy, where maybe 50 are useful, and then the time used for eliminating 9950 of the templates explicitly would start to matter. So for longer runs on tougher to predict compounds, it leads to a higher success rate of finding routes as there are more options for the tree search, but for the majority of compounds it does not give faster search times to first route found (see Figure 9 in the MLST paper). If the lower priority, but applicable, templates are then merely applicable in silico and not chemically relevant, is something that's a bit tougher to measure, but we are working on that aspect.

Get it. Thanks for your enlightening reply.
